src/arch/riscv/kernel/vspace.c

```diff
@@ -70,10 +70,10 @@ static pte_t pte_next(word_t phys_addr, bool_t is_leaf)
 
     return pte_new(ppn,
                    0,     /* sw */
-                   1,     /* dirty */
-                   1,     /* accessed */
+                   is_leaf ? 1 : 0,     /* dirty leaf/reserved non-leaf */
+                   is_leaf ? 1 : 0,     /* accessed leaf/reserved non-leaf */
```

Comment:
Does anyone know why these were set to 1 in the first place? We don't usually used dirty/accessed in the kernel.

Reply:
That was done by @heshamelmatary in the initial port. So maybe it has no particular reason as the spec was not saying much.

Reply:
For leaf tables, they have to be set to 1 if you want the OS to ignore them.

The privileged spec allows two implementations:

- When a virtual page is accessed and the A bit is clear, or is written and the D bit is clear, a
  page-fault exception is raised.
- When a virtual page is accessed and the A bit is clear, or is written and the D bit is clear, the
  implementation sets the corresponding bit(s) in the PTE. The PTE update must be atomic
  with respect to other accesses to the PTE, and must atomically check that the PTE is valid
  and grants sufficient permissions. Updates of the A bit may be performed as a result of
  speculation, but updates to the D bit must be exact (i.e., not speculative), and observed in
  program order by the local hart. Furthermore, the PTE update must appear in the global
  memory order no later than the explicit memory access, or any subsequent explicit memory
  access to that virtual page by the local hart. The ordering on loads and stores provided by
  FENCE instructions and the acquire/release bits on atomic instructions also orders the PTE
  updates associated with those loads and stores as observed by remote harts.
  The PTE update is not required to be atomic with respect to the explicit memory access that
  caused the update, and the sequence is interruptible. However, the hart must not perform
  the explicit memory access before the PTE update is globally visible.

Ideally, seL4 would expose these to user space but it seems that this is not currently the case.

Reply:
Thanks for looking these up.

Ok, this means we do want to set them to 1 for now. Exporting the mechanism to user space would be an option, but it would be an API change that we should do separately if we decide to do it, and it should go through an RFC.

Reply:
We always want to leave them as 1s for the kernel, for leaf pages.

WRT user pages, I didn't see a good reason to expose D/A to userspace back then as no project/lib (even on Arm and x86) needed to make use of them for page placement policies; seL4/userspace don't do (demand) paging/swapping to disk anyway.

Reply:
I think the primary benefit of exposing it to user space will be for people writing dynamic systems in the future. I agree that there are currently no systems that use seL4 in that way, but I'd certainly like there to be.

Reply:
The UNSW AOS course has students implement demand paging :)